### ISSUE TITLE
E0446 Doesn't Trigger For Types In Private Modules

### ISSUE THREAD

**Issue by danii** [rust-lang/rust#89264]
Declaring a public type and a public function in a private module, and then publicly reexporting the function doesn't trigger the E0446 error on using private types in public interfaces.

```rust
mod module {
  pub struct Private; // This isn't available outside of the crate because the module it is in is private.
  
  pub fn public() -> Private {Private} // This is available outside of the crate...
}

pub use module::public; // ...because of this reexport here.
```

I expected to see the E0446 error, because the above code is semantically equivalent to this code, from a downstream crate's point of view.
```rust
struct Private;

pub fn public() -> Private {Private} // error[E0446]: private type `Private2` in public interface
```

### Meta
This bug occurs both on Rust version `rustc 1.56.0-nightly (3d0774d0d 2021-08-18)` and version `rustc 1.54.0 (a178d0322 2021-07-26)`.

**Comment by ExpHP** [comment]
I don't think this is a bug.  It is a central part of the [sealed trait](https://rust-lang.github.io/api-guidelines/future-proofing.html) pattern.
